### PR TITLE
implements RFC 463; produce gcc compatible dep files

### LINKED ISSUE #463: compiler: Option to produce a depfile from target dependencies
### Title

Depfile

### Abstract

Add a new command-line option (e.g. `--depfile`) for the `nim compile` command. It accepts an argument (output path for a depfile, generated for a given target).

### Motivation

This change would improve Nim integrations with language-independent build
systems. Particularly, get better edit-compile cycle times.

Currently a build system needs to rebuild a binary if _any_ of source files has
been changed, even if they aren't used by the binary. The full list of files
that a given source file depends on can only be discovered by the compiler.

It doesn't need to be a standalone command because if the file has never been
compiled, it must be built anyway, generating the header dependencies as a side
effect.

### Description

**Downside:** growing number of command-line flags :)

**Similar proposal:** #412. However custom JSON format will not be understood by
build systems.

### Depfile file format

From GCC manual (`-M` option):

> The preprocessor outputs one 'make' rule containing the object file name for
> that source file, a colon, and the names of all the included files, including
> those coming from '-include' or '-imacros' command-line options.

Or, in the BNF notation:  
https://cmake.org/cmake/help/latest/command/add_custom_command.html

It is recognized by all common build systems.

### Before
*(no command to produce a depfile)*

### After
```sh
$ nim c --depfile main.d main
$ head -4 main.d # system imports not shown
main: \
	/home/project/main.nim \
	/home/project/src/file1.nim \
	/home/project/src/file2.nim \
```

#### Use the depfile in Makefile

```Makefile
main: main.nim
	nim c $(NIMFLAGS) --depfile $@.d $<

-include main.d
```

#### Use the depfile in Ninja

```Ninja
rule nimc:
  deps = gcc
  depfile = $out.d
  command = nim c $nimflags --depfile $out.d $in

build main: nimc main.nim
```

#### Use the depfile in CMake

```CMake
...

add_custom_command(
    OUTPUT main
    COMMAND nim c ${NIMFLAGS} --depfile main.d main.nim
    DEPFILE main.d
    VERBATIM
)
```

#### Use the depfile in Meson

```Meson
...

nim = find_program('nim')
custom_target(
  input : 'main.nim',
  output : 'main',
  depfile : '@BASENAME@.d',
  command : [nim, 'c', '--depfile', '@DEPFILE@', '@INPUT@'],
  build_by_default : true
)
```

### Backwards Compatibility

This RFC introduces no backward-incompatible changes.

### PR DESCRIPTION
**Pull request by ringabout**
closes https://github.com/nim-lang/RFCs/issues/463

todo

- [x] changelog
- [x] better error messages
- [x] allow specifying the nim executable 
- [x] allow specifying nimcache
- [x] help messages

### CODE REVIEW

**Comment by CyberTailor** [comment]
Could you [rebase](https://git-rebase.io/) commits for cleaner logs?

**Comment by CyberTailor** [comment]
So `nimgendeps` is a wrapper around `nim c`? Very clever!

**Comment by CyberTailor** [comment]
Also I'd:

1) pass all unknown options to Nim
2) add "--nim" option to set Nim executable
2) respect "--nimcache" option
3) add help output

**Comment by CyberTailor** [comment]
Depfile should be written not for "src" but for out binary.

File: changelog.md [hunk 1]
```diff
@@ -140,3 +140,8 @@
   a simple configuration change setting that you can do yourself too. In `$nim/config/nim.cfg`
   replace `pkgs` by `pkgs2`.
 
+
+- There is a new switch `--nimMainPrefix:prefix` to influence the `NimMain` that the
+  compiler produces. This is particularly useful for generating static libraries.
```
Comment (ringabout) [inline changelog.md]:
```suggestion
```

File: changelog.md [hunk 2]
```diff
@@ -140,3 +140,6 @@
   a simple configuration change setting that you can do yourself too. In `$nim/config/nim.cfg`
   replace `pkgs` by `pkgs2`.
 
+
+
```
Comment (ringabout) [inline changelog.md]:
```suggestion
```

**Comment by ringabout** [comment]
> 1) pass all unknown options to Nim
> 2) add "--nim" option to set Nim executable
> 3) respect "--nimcache" option
> 4) add help output
> 5) Depfile should be written not for "src" but for out binary.

@CyberTailor Thanks for your suggestions. The points above are solved.

File: tools/nimgendeps.nim [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    let target = src.changeFileExt(when defined(windows): ".exe" else: "").handleMakeEscape()
+    f.write(target & ": \\" & '\n')
+
+    for i in 0 ..< deps.len-1:
+      let path = deps[i][0].getStr.handleMakeEscape()
+      f.write('\t' & path & " \\" & '\n')
+    let path = deps[^1][0].getStr.handleMakeEscape()
+    f.write('\t' & path)
+  finally:
+    f.close()
```
Comment (Varriount) [inline tools/nimgendeps.nim:59]: Would this work instead?
```suggestion
  if option.nimcacheDir.len == 0:
    option.nimcacheDir = getTempDir() / "nimgccdeps"
  if option.passNim.len == 0:
    option.passNim = "c"
  if option.nimBinary.len == 0:
    option.nimBinary = "nim"

  let (msg, exitCode) = execCmdEx(
    fmt"{option.nimBinary} {option.passNim} --d:nimBetterRun --nimcache:{option.nimcacheDir} {src}"
  )
  doAssert exitCode == 0, msg
  
  let jsonFileName = changeFileExt(extractFileName(src), ".json")
  let jsonFilePath = option.nimcacheDir / jsonFile
  let jsonFileData = parseJson(readFile(jsonFilePath))
  let depFiles = jsonFileData["depfiles"]

  var extension: string
  if defined(windows):
    extension = ".exe"
  else:
    extension = ""
  
  var outFile: File
  if option.outFile.len == 0:
    outFile = stdout
  else:
    outFile = open(option.outFile, fmWrite)

  try:
    let target = changeFileExt(src, extension)
    outFile.write(handleMakeEscape(target))
    outFile.write(": ")

    for i in 0 ..< depFiles.len:
      let path = getStr(depFiles[i][0])
      outfile.write("\\\n")
      outFile.write('\t')
      outFile.write(handleMakeEscape(path))

    outfile.write("\n")
  finally:
    f.close()
```

File: tools/nimgendeps.nim [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      src = p.key
+      break
+
+  if isStop:
+    discard
+  elif src.len == 0:
+    echo helpText
+  else:
+    writeDepfile(src, DepOption(passNim: passNim, outFile: outFile,
+                  nimBinary: nimBinary, nimcacheDir: nimcacheDir))
```
Comment (Varriount) [inline tools/nimgendeps.nim:103]:
```suggestion
  var passNim = ""
  var outFile = ""
  var sourceFile = ""
  var nimBinary = ""
  var nimcacheDir = ""
  var argumentError = false

  let parser = initOptParser()
  while true:
    parser.next()

    case parser.kind
    of cmdEnd:
      break

    of cmdLongOption:
      if cmpIgnoreStyle(parser.key, "passNim") == 0:
        passNim = parser.val
      elif cmpIgnoreStyle(parser.key, "out") == 0:
        outFile = parser.val
      elif cmpIgnoreStyle(parser.key, "nim") == 0:
        nimBinary = parser.val
      elif cmpIgnoreStyle(parser.key, "nimcache") == 0:
        nimcacheDir = parser.val
      else:
        echo "unexpected option: ", parser.key
        argumentError = true

    of cmdShortOption:
      if parser.key == "o":
        outFile = parser.val
      else:
        echo "unexpected option: ", parser.key
        argumentError = true

    of cmdArgument:
      if sourceFile.len == 0:
        sourceFile = parser.key
      else:
        echo "unexpected argument: ", parser.key
        argumentError = true

  if argumentError:
    discard

  elif sourceFile.len == 0:
    echo helpText
  else:
    writeDepfile(
      src,
      DepOption(
        passNim: passNim,
        outFile: outFile,
        nimBinary: nimBinary,
        nimcacheDir: nimcacheDir
      )
    )
```

**Comment by Varriount** [comment]
@ringabout Please see comments

**Pull request closed without merging**